photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/search/RandomSearch.scala

```diff
@@ -184,16 +160,21 @@ class RandomSearch[T](
    */
   protected[search] def drawCandidates(n: Int): DenseMatrix[Double] = {
     // Draw candidates from a Sobol generator, which produces values in the range [0, 1]
-    val candidates = (1 until n).foldLeft(DenseMatrix(paramDistributions.nextVector)) { case (acc, _) =>
+    (1 until n).foldLeft(DenseMatrix(paramDistributions.nextVector)) { case (acc, _) =>
       DenseMatrix.vertcat(acc, DenseMatrix(paramDistributions.nextVector))
     }
+  }
 
-    // Adjust candidates according to specified ranges
-    ranges.zipWithIndex.foreach { case (range, j) =>
-      candidates(::,j) *= (range.end - range.start)
-      candidates(::,j) += range.start
+  /**
+   * Discretize candidates whose values are integers after rescaling
+   *
+   * @param candidate candidate with values in [0, 1]
+   * @param discreteParams Map that specifies the indices of discrete parameters and their numbers of discrete values
+   *
+   */
+  protected [search] def discretizeCandidate(candidate: DenseVector[Double], discreteParams: Map[Int, Int]): Unit = {
+    discreteParams.foreach { case (index, numDiscreteValues) =>
+      candidate(index) = floor(candidate(index) * numDiscreteValues) / numDiscreteValues
```

Comment:
Wont this throw an error since candidate it immutable?

Reply:
I think you can't resign another DenseVector to candidate becasue the object reference can not be accessed. But modify the element in a DenseVector is fine.

Reply:
@cmjiang is right -- `DenseVector` is a mutable object.

@basukinjal I'd used `floor` for descretizing in the past, and you asked me to change it to `round`. Do you remember why? I think you were worried about down-biasing candidates.